srml/support/src/lib.rs

```diff
@@ -1,5 +1,176 @@
+where
+	Balance: Clone + Copy + Default + SimpleArithmetic,
+	Moment: Clone + Copy + PartialOrd + SaturatedConversion + rstd::convert::TryInto<u64>,
+{
+	pub fn can_withdraw(&self, at: Moment, new_balance: Balance) -> bool {
+		match self {
+			DetailLock::BalanceDetailLock(lock) => lock.can_withdraw(at, new_balance),
+			DetailLock::StakingAndUnbondingDetailLock(lock) => {
+				lock.can_withdraw(at.saturated_into::<TimeStamp>(), new_balance)
+			}
+		}
+	}
+}
+
+//impl<Balance, Moment> PartialEq for DetailLock<Balance, Moment>
+//where
+//	Balance: PartialEq,
+//	Moment: PartialEq,
+//{
+//	#[inline]
+//	fn eq(&self, other: &Self) -> bool {
+//		match (self, other) {
+//			(DetailLock::BalanceDetailLock(a), DetailLock::BalanceDetailLock(b)) => a == b,
+//			(DetailLock::StakingAndUnbondingDetailLock(a), DetailLock::StakingAndUnbondingDetailLock(b)) => a == b,
+//			_ => false,
+//		}
+//	}
+//}
+
+#[derive(Encode, Decode, Clone, PartialEq, Eq, RuntimeDebug)]
+pub struct BalanceLock<Balance, Moment> {
+	pub amount: Balance,
+	pub until: Moment,
+}
+
+impl<Balance, Moment> BalanceLock<Balance, Moment>
+where
+	Balance: Clone + Copy + Default + SimpleArithmetic,
+	Moment: Clone + Copy + PartialOrd,
+{
+	fn can_withdraw(&self, at: Moment, new_balance: Balance) -> bool {
+		self.until > at && self.amount > new_balance
+	}
+}
+
+#[derive(PartialEq, Eq, Clone, Encode, Decode, RuntimeDebug)]
+pub struct UnlockChunk<Balance, Moment> {
+	/// Amount of funds to be unlocked.
+	// TODO: Compact?
+	pub value: Balance,
+	/// Era number at which point it'll be unlocked.
+	pub until: Moment,
+}
+
+#[derive(Clone, PartialEq, Eq, Encode, Decode, RuntimeDebug)]
+pub struct StakingAndUnbondingLock<Balance, Moment> {
+	pub staking_amount: Balance,
+	// now >= l.until || new_balance >= l.amount
+	/// Any balance that is becoming free, which may eventually be transferred out
+	/// of the stash (assuming it doesn't get slashed first).
+	pub unlocking: Vec<UnlockChunk<Balance, Moment>>,
+}
+
+impl<Balance: Default, Moment: Default> Default for StakingAndUnbondingLock<Balance, Moment> {
+	fn default() -> Self {
+		StakingAndUnbondingLock {
+			staking_amount: Default::default(),
+			unlocking: Default::default(),
+		}
+	}
+}
+
+//impl<Balance, Moment> PartialEq for StakingAndUnbondingLock<Balance, Moment>
+////where
+////	Balance: PartialEq,
+////	Moment: PartialEq,
+//{
+//	#[inline]
+//	fn eq(&self, other: &Self) -> bool {
+//		self.staking_amount == other.staking_amount
+//			&& self.unlocking.len() == other.unlocking.len()
+//			&& self.unlocking.iter().zip(other.unlocking.iter()).all(|x, y| x == y)
+//	}
+//}
+
+impl<Balance, Moment> StakingAndUnbondingLock<Balance, Moment>
+where
+	Balance: Clone + Copy + Default + SimpleArithmetic,
+	Moment: Clone + Copy + PartialOrd,
+{
+	fn can_withdraw(&self, at: Moment, new_balance: Balance) -> bool {
+		// TODO: Is it correct to use clone here?
+		let mut locked_amount = self.staking_amount.clone();
```

Comment:
`staking_amount` has a `Copy` bound. If we want to take its ownership, it'll be cloned automactically.